xla/service/gpu/transforms/block_scaling_rewriter.cc

```diff
@@ -194,39 +196,199 @@ absl::StatusOr<HloInstruction*> ExpandDequantizeCustomCall(
+      block_size == BlockScalingRewriter::kBlockSizeMXFP8) {
+    return CompositeType::MXFP8_E5M2;
+  }
 
-// Build HLO for scaled dot op.
-absl::StatusOr<XlaOp> BuildBlockScaledDot(XlaBuilder& builder,
-                                          const HloInstruction* lhs_input,
-                                          const HloInstruction* rhs_input,
-                                          const HloInstruction* lhs_scale,
-                                          const HloInstruction* rhs_scale,
+  return CompositeType::CUSTOM_TYPE;
+}
+
+bool IsSupportedByCudnn(CompositeType lhs, CompositeType rhs) {
+  // cuDNN supports mixing input types for MXFP8, but the E5M2/E5M2 combination
+  // is not supported.
+  return (lhs == CompositeType::MXFP8_E4M3FN &&
+          rhs == CompositeType::MXFP8_E4M3FN) ||
+         (lhs == CompositeType::MXFP8_E4M3FN &&
+          rhs == CompositeType::MXFP8_E5M2) ||
+         (lhs == CompositeType::MXFP8_E5M2 &&
+          rhs == CompositeType::MXFP8_E4M3FN);
+}
+
+// Reshape inputs to shapes compatible with cuDNN.
+absl::StatusOr<std::tuple<XlaOp, XlaOp, int64_t>> BuildScaledDotInputs(
+    XlaOp input_op, XlaOp scale_op) {
+  // Get shapes from the inputs.
+  XlaBuilder& builder = *input_op.builder();
+  TF_ASSIGN_OR_RETURN(Shape input_shape, builder.GetShape(input_op));
+  TF_ASSIGN_OR_RETURN(Shape scale_shape, builder.GetShape(scale_op));
+  TF_RET_CHECK(input_shape.rank() == 2 || input_shape.rank() == 3);
+
+  // Calculate output shape size.
+  int64_t batch_size = input_shape.rank() == 3 ? input_shape.dimensions(0) : 1;
+  int64_t size_contracting = input_shape.dimensions().back();
+  int64_t size_noncontracting = input_shape.dimensions(input_shape.rank() - 2);
+  int64_t scale_contracting = scale_shape.dimensions().back();
+
+  // Reshape inputs, if necessary.
+  if (input_shape.rank() != 3) {
+    input_op = Reshape(input_op, {1, size_noncontracting, size_contracting});
+    scale_op = Reshape(scale_op, {1, size_noncontracting, scale_contracting});
+  }
+
+  // cuDNN kernel imposes constraints on the input shape sizes.
+  const int64_t kInputNonContractingTileSize = 128;
+  const int64_t kScaleContractingTileSize = 4;
+
+  // Pad inputs, if necessary.
+  if (size_noncontracting % kInputNonContractingTileSize != 0 ||
+      scale_contracting % kScaleContractingTileSize != 0) {
+    // Calculate new output shape sizes.
+    int64_t padded_noncontracting =
+        RoundUpTo(size_noncontracting, kInputNonContractingTileSize);
+    int64_t padded_contracting =
+        RoundUpTo(scale_contracting, kScaleContractingTileSize);
+
+    // Build padding configs.
+    PaddingConfig input_padding_config = MakeNoPaddingConfig(/*rank=*/3);
+    if (size_noncontracting != padded_noncontracting) {
+      input_padding_config.mutable_dimensions(1)->set_edge_padding_high(
+          padded_noncontracting - size_noncontracting);
+    }
+    PaddingConfig scale_padding_config = input_padding_config;
```

Comment:
Why is padding for the input the same as scale?

Suppose we have input dimension size 253 (so padding is 3 to pad to multiple of 4), and scaled block size is 64 (so the scales dimension is 4) -- we don't want to pad 3 to the scales dimension, do we?

Reply:
Ah, just realized it's for non-contracting dimension, which is the same.